src/python/pants/engine/internals/parametrize.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import itertools
+from dataclasses import dataclass
+from typing import Any, Iterator
+
+from pants.engine.addresses import Address
+from pants.util.meta import frozen_after_init
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class Parametrize:
+    """A builtin function/dataclass that can be used to parametrize Targets.
+
+    Parameterization is applied between TargetAdaptor construction and Target instantiation, which
+    means that individual Field instances need not be aware of it.
+    """
+
+    args: tuple[str, ...]
+    kwargs: dict[str, Any]
```

Comment:
I'd expect a `FrozenDict[str, ImmutableValue]` as the type for `kwargs` here. (Where `ImmutableValue` simply is `Any` but with a name that suggests it needs to be a immutable/hashable value, similarly done in `target.py`)

Reply:
So, the phase at which we're operating here is before fields have been parsed into their final values: we don't yet know which target type they will later be applied to, so we don't know what field type they correspond. At this point, the values can be anything (`list` is quite common for target arguments, for example).

So `parametrize` floats around in a few `fields: dict[str, Any]` lists until it is expanded. Given that, there isn't reason to go through freezing. If that changes at some point in the future, can adapt though.